### PR TITLE
KAFKA-7734: Metrics tags should use LinkedHashMap to guarantee ordering

### PR DESCRIPTION
**Pull request by lambdaliu**
This pull request replaces HashMap with LinkedHashMap to guarantee ordering of metrics tags.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by vvcephei** [comment]
Thanks for the PR, @lambdaliu !

I realize `LinkedHashMap` is specified in the ticket, but would `TreeMap` be better? Then, we could rely on tags being alphanumerically sorted in the documentation without having to painstakingly manage the order of insertion in the code.

**Comment by vvcephei** [comment]
Also, if we were to go with `TreeMap`, then this PR could be a one-line change in `MetricName`, rather than having to change every spot in the code where we create a tag map.

**Comment by ijuma** [comment]
We should consider the memory cost of changing MetricName too.

**Comment by ijuma** [comment]
That is, sorting it once and storing it sorted should be considered too.

**Comment by guozhangwang** [comment]
> Also, if we were to go with `TreeMap`, then this PR could be a one-line change in `MetricName`, rather than having to change every spot in the code where we create a tag map.

> That is, sorting it once and storing it sorted should be considered too.

+1, good idea!

**Comment by vvcephei** [comment]
Thanks for that point, @ijuma.

It's true that we could also sort the tags and store them in a LinkedHashMap in the MetricName constructor.

I'm not too familiar with the memory overhead of the different data structures. What did you have in mind regarding the memory footprint of HashMap (current) vs LinkedHashMap or TreeMap?

From my surface-level knowledge, it seems like the LinkedHashMap needs to store a couple of extra pointers for each element, but is otherwise similar. TreeMap needs to store a few pointers as well as the bookkeeping info for balancing the tree. None of this jumps out as significant, but it's also a shallow analysis.

I suppose that it might be worth analyzing the cost to access the tags. I'm guessing they're overwhelmingly iterated instead of used for key lookups, which makes both HashMap and LinkedHashMap just linear, whereas the TreeMap would be linear plus the `log(n)` lookup to find the min node. Of course, since we have a small number of tags, `log(n)` is practically `1` anyway...

If we wish to speed up creating the JMX bean names, we could circumvent the whole question by generating the bean name once and caching it for each MethodName.

Can you elaborate on your concern?

Thanks,
-John

**Comment by lambdaliu** [comment]
Thank you for the comments, @vvcephei 
If we change the order of the metrics tags, most MBeanName in the document will change. This will cause some Kafka monitoring tools to fail to get the metrics, such as Kafka manager. So how to keep the compatibility is a problem.

**Review by guozhangwang — COMMENTED** [review]
@lambdaliu Thanks for the PR! I've made a pass over it and here are some meta comments:

1. I agree with you that we should maintain compatibility, and hence we'd avoid sorting the tag keys alphabetically, but just use LinkedHashMap instead. Footprint wise, I would not worry too much with `LinkedHashMap` v.s. `HashMap` for the reason brought up by @vvcephei.

2. There may be other places that also needs to be updated. For example: `StreamsMetricsImpl#tagMap`. I made a quick pass and it seems this is the only other place, but maybe you could make another pass just to double check all the callers of the metrics tag generation and see if there's any other places that we'd need to update.

In addition to this, I'd suggest we add a unit test for module's metrics that checks the corresponding generated mbean names by the jmx reporter (just representing any reporter that may rely on metric.tags().keySet()) has the expected string concatenation. If you want to keep this PR small we can do it in a follow-up one.

**Comment by vvcephei** [comment]
Sounds good @lambdaliu . Thanks for considering it!

**Comment by ijuma** [comment]
Looking at the PR, we didn't guarantee ordering before, right? This makes sense because tags and ordering is a bit weird. The main issue is that the JMX reporter basically expands the tags as the JMX metric name. It seems to me that we need a KIP for this.

**Comment by guozhangwang** [comment]
@ijuma Could you elaborate a bit why we need a KIP for this? This PR is not changing the ordering of the tags.

**Comment by ijuma** [comment]
So why do tests have to be changed?

**Comment by guozhangwang** [comment]
@ijuma I've just made the single non-test change and run unit tests un-changed, they still pass. I think @lambdaliu modified the unit test just for mechanism consistency, but I'd leave it to himself to explain.

**Comment by lambdaliu** [comment]
Hi @guozhangwang, thanks for reviewing. I will check again later. And for the unit test, I will do it in a follow-up one PR.

**Comment by lambdaliu** [comment]
Hi @ijuma, just like @guozhangwang said, I change the unit test just for consistency.

**Comment by ijuma** [comment]
One thing to keep in mind is is that if you use the default `Map.apply` method, you get one of the following if the map has 4 items or less and the order is preserved. So, using `LinkedHashMap` just makes it less efficient.

```scala
  @SerialVersionUID(-5626373049574850357L)
  private object EmptyMap extends AbstractMap[Any, Nothing] with Map[Any, Nothing] with Serializable {
    override def size: Int = 0
    override def apply(key: Any) = throw new NoSuchElementException("key not found: " + key)
    override def contains(key: Any) = false
    def get(key: Any): Option[Nothing] = None
    override def getOrElse [V1](key: Any, default: => V1): V1 = default
    def iterator: Iterator[(Any, Nothing)] = Iterator.empty
    override def updated [V1] (key: Any, value: V1): Map[Any, V1] = new Map1(key, value)
    def + [V1](kv: (Any, V1)): Map[Any, V1] = updated(kv._1, kv._2)
    def - (key: Any): Map[Any, Nothing] = this
  }

  @SerialVersionUID(-9131943191104946031L)
  class Map1[K, +V](key1: K, value1: V) extends AbstractMap[K, V] with Map[K, V] with Serializable {
    override def size = 1
    override def apply(key: K) = if (key == key1) value1 else throw new NoSuchElementException("key not found: " + key)
    override def contains(key: K) = key == key1
    def get(key: K): Option[V] =
      if (key == key1) Some(value1) else None
    override def getOrElse [V1 >: V](key: K, default: => V1): V1 =
      if (key == key1) value1 else default
    def iterator = Iterator((key1, value1))
    override def updated [V1 >: V] (key: K, value: V1): Map[K, V1] =
      if (key == key1) new Map1(key1, value)
      else new Map2(key1, value1, key, value)
    def + [V1 >: V](kv: (K, V1)): Map[K, V1] = updated(kv._1, kv._2)
    def - (key: K): Map[K, V] =
      if (key == key1) Map.empty else this
    override def foreach[U](f: ((K, V)) => U): Unit = {
      f((key1, value1))
    }
  }

  @SerialVersionUID(-85684685400398742L)
  class Map2[K, +V](key1: K, value1: V, key2: K, value2: V) extends AbstractMap[K, V] with Map[K, V] with Serializable {
    override def size = 2
    override def apply(key: K) =
      if (key == key1) value1
      else if (key == key2) value2
      else throw new NoSuchElementException("key not found: " + key)
    override def contains(key: K) = (key == key1) || (key == key2)
    def get(key: K): Option[V] =
      if (key == key1) Some(value1)
      else if (key == key2) Some(value2)
      else None
    override def getOrElse [V1 >: V](key: K, default: => V1): V1 =
      if (key == key1) value1
      else if (key == key2) value2
      else default
    def iterator = Iterator((key1, value1), (key2, value2))
    override def updated [V1 >: V] (key: K, value: V1): Map[K, V1] =
      if (key == key1) new Map2(key1, value, key2, value2)
      else if (key == key2) new Map2(key1, value1, key2, value)
      else new Map3(key1, value1, key2, value2, key, value)
    def + [V1 >: V](kv: (K, V1)): Map[K, V1] = updated(kv._1, kv._2)
    def - (key: K): Map[K, V] =
      if (key == key1) new Map1(key2, value2)
      else if (key == key2) new Map1(key1, value1)
      else this
    override def foreach[U](f: ((K, V)) => U): Unit = {
      f((key1, value1)); f((key2, value2))
    }
  }

  @SerialVersionUID(-6400718707310517135L)
  class Map3[K, +V](key1: K, value1: V, key2: K, value2: V, key3: K, value3: V) extends AbstractMap[K, V] with Map[K, V] with Serializable {
    override def size = 3
    override def apply(key: K) =
      if (key == key1) value1
      else if (key == key2) value2
      else if (key == key3) value3
      else throw new NoSuchElementException("key not found: " + key)
    override def contains(key: K) = (key == key1) || (key == key2) || (key == key3)
    def get(key: K): Option[V] =
      if (key == key1) Some(value1)
      else if (key == key2) Some(value2)
      else if (key == key3) Some(value3)
      else None
    override def getOrElse [V1 >: V](key: K, default: => V1): V1 =
      if (key == key1) value1
      else if (key == key2) value2
      else if (key == key3) value3
      else default
    def iterator = Iterator((key1, value1), (key2, value2), (key3, value3))
    override def updated [V1 >: V] (key: K, value: V1): Map[K, V1] =
      if (key == key1)      new Map3(key1, value, key2, value2, key3, value3)
      else if (key == key2) new Map3(key1, value1, key2, value, key3, value3)
      else if (key == key3) new Map3(key1, value1, key2, value2, key3, value)
      else new Map4(key1, value1, key2, value2, key3, value3, key, value)
    def + [V1 >: V](kv: (K, V1)): Map[K, V1] = updated(kv._1, kv._2)
    def - (key: K): Map[K, V] =
      if (key == key1)      new Map2(key2, value2, key3, value3)
      else if (key == key2) new Map2(key1, value1, key3, value3)
      else if (key == key3) new Map2(key1, value1, key2, value2)
      else this
    override def foreach[U](f: ((K, V)) => U): Unit = {
      f((key1, value1)); f((key2, value2)); f((key3, value3))
    }
  }

  @SerialVersionUID(-7992135791595275193L)
  class Map4[K, +V](key1: K, value1: V, key2: K, value2: V, key3: K, value3: V, key4: K, value4: V) extends AbstractMap[K, V] with Map[K, V] with Serializable {
    override def size = 4
    override def apply(key: K) =
      if (key == key1) value1
      else if (key == key2) value2
      else if (key == key3) value3
      else if (key == key4) value4
      else throw new NoSuchElementException("key not found: " + key)
    override def contains(key: K) = (key == key1) || (key == key2) || (key == key3) || (key == key4)
    def get(key: K): Option[V] =
      if (key == key1) Some(value1)
      else if (key == key2) Some(value2)
      else if (key == key3) Some(value3)
      else if (key == key4) Some(value4)
      else None
    override def getOrElse [V1 >: V](key: K, default: => V1): V1 =
      if (key == key1) value1
      else if (key == key2) value2
      else if (key == key3) value3
      else if (key == key4) value4
      else default
    def iterator = Iterator((key1, value1), (key2, value2), (key3, value3), (key4, value4))
    override def updated [V1 >: V] (key: K, value: V1): Map[K, V1] =
      if (key == key1)      new Map4(key1, value, key2, value2, key3, value3, key4, value4)
      else if (key == key2) new Map4(key1, value1, key2, value, key3, value3, key4, value4)
      else if (key == key3) new Map4(key1, value1, key2, value2, key3, value, key4, value4)
      else if (key == key4) new Map4(key1, value1, key2, value2, key3, value3, key4, value)
      else (new HashMap).updated(key1,value1).updated(key2, value2).updated(key3, value3).updated(key4, value4).updated(key, value)
    def + [V1 >: V](kv: (K, V1)): Map[K, V1] = updated(kv._1, kv._2)
    def - (key: K): Map[K, V] =
      if (key == key1)      new Map3(key2, value2, key3, value3, key4, value4)
      else if (key == key2) new Map3(key1, value1, key3, value3, key4, value4)
      else if (key == key3) new Map3(key1, value1, key2, value2, key4, value4)
      else if (key == key4) new Map3(key1, value1, key2, value2, key3, value3)
      else this
    override def foreach[U](f: ((K, V)) => U): Unit = {
      f((key1, value1)); f((key2, value2)); f((key3, value3)); f((key4, value4))
    }
  }
```

**Comment by lambdaliu** [comment]
retest this please

**Comment by lambdaliu** [comment]
retest this please

**Comment by guozhangwang** [comment]
> One thing to keep in mind is is that if you use the default `Map.apply` method

Not sure if I understand? this PR is for updating the java code only.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/NamedCache.java [hunk 1]
```diff
@@ -367,8 +367,8 @@ private NamedCacheMetrics(final StreamsMetricsImpl metrics, final String cacheNa
 
             // add parent
             final Map<String, String> allMetricTags = metrics.tagMap(
-                "record-cache-id", "all",
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/NamedCache.java:370]: Nice catch.

**Comment by ijuma** [comment]
@guozhangwang the original commit had changed many Scala files.

**Comment by guozhangwang** [comment]
> @guozhangwang the original commit had changed many Scala files.

Ack!

**Comment by guozhangwang** [comment]
@lambdaliu Thanks for the newest patch! LGTM and merging to trunk.

**Comment by guozhangwang** [comment]
Also cherry-picked to 2.1

**Pull request merged**